Migrate Privacy Policy system from SMF 2.0.16

MisterVector opened the pull request:
This pull request migrates the privacy policy system added to SMF 2.0.16.

This closes #6075

MisterVector committed:
Migrate Privacy Policy system from SMF 2.0.16

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

MisterVector committed:
Re-added some things that went missing and other changes

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

MisterVector commented:
I may have missed some things as I'm not very familiar with the codebase. Just point me in the right direction!

SychO9 commented on Themes/default/Register.template.php:
```diff
@@ -701,4 +727,73 @@ function template_edit_reserved_words()
+						<input type="submit" value="', $txt['save'], '" tabindex="', $context['tabindex']++, '" class="button" onclick="return resetPolicyConfirm()" />
+						<input type="hidden" name="policy_lang" value="', $context['current_policy_lang'], '" />
+						<input type="hidden" name="sa" value="policy" />
+						<input type="hidden" name="', $context['session_var'], '" value="', $context['session_id'], '" />
```
Need to also add the token you created in the source files, `admin-regp`

a hidden input containing the token details
```php
<input type="hidden" name="', $context['admin-regp_token_var'], '" value="', $context['admin-regp_token'], '" />
```

MisterVector committed:
Added missing hidden input tag that contains the admin-regp token

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

sbulen commented:
Getting a couple undefineds - I think this was from editing the policies:
![image](https://user-images.githubusercontent.com/23568484/80935594-2be64e80-8d82-11ea-8fd9-e6ee7ffd3545.png)

XinYenFon commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,91 @@
+		if ($context['can_accept_agreement'])
+		{
+			echo '
+		<p class="description">
```
we are no longer using p element as description, please use div instead.

XinYenFon commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,91 @@
+		elseif (!empty($context['agreement_accepted_date']))
+		{
+			echo '
+		<p class="description">
```
same here p > div

XinYenFon commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,91 @@
+	if (!empty($context['policy']))
+	{
+		if (!empty($context['agreement']))
+			echo '<br />';
```
Invalid HTML5 :) you can leave this br tag out completely because block elements and proper padding/margin should be enough to create space.

XinYenFon commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,91 @@
+		if ($context['can_accept_privacy_policy'])
+		{
+			echo '
+		<p class="description">
```
p > div

XinYenFon commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,91 @@
+		elseif (!empty($context['privacy_policy_accepted_date']))
+		{
+			echo '
+		<p class="description">
```
p > div

XinYenFon commented on Themes/default/Register.template.php:
```diff
@@ -701,4 +727,74 @@ function template_edit_reserved_words()
+					</div>
+				</form>
+			</div>
+			<span class="botslice"><span></span></span>
```
slices are no longer used in SMF 2.1

XinYenFon commented on Themes/default/Register.template.php:
```diff
@@ -701,4 +727,74 @@ function template_edit_reserved_words()
+			</div>
+			<span class="botslice"><span></span></span>
+		</div>
+		<br class="clear" />';
```
this should be unnecessary as well at this stage

MisterVector replied:
Actually, this appears in a lot of other places in the SMF 2.1 files (without the trailing slash). Not sure I should touch those as they don't fall within the scope of this PR.

SychO9 reviewed (commented):
I also believe the `requirePolicyAgreement` needs to be set to 1 by default on installation, i.e make sure the value is inserted to the settings table in the sql files `install_2-1_mysql` and `install_2-1_postgresql` just like `requireAgreement`

MisterVector committed:
p -> div and removed other unnecessary elements

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

XinYenFon commented:
You forgot the closing tags ,🙂

MisterVector committed:
Whoops!

Close the equivalent tags! And removed something else I did not
mean to add in this commit

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

SychO9 commented on Sources/ManageRegistration.php:
```diff
@@ -292,20 +322,33 @@ function SetReserved()
+		array('check', 'requirePolicyAgreement', 'text_label' => $txt['admin_privacy_policy'], 'value' => 1, 'disabled' => 1),
+		array('warning', empty($policy) ? 'error_no_privacy_policy' : ''),
+	'',
+		array('int', 'coppaAge', 'subtext' => $txt['setting_coppaAge_desc'], 'onchange' => 'checkCoppa();'),
```
Instead of creating a new text string, please preserve the previous coppaAge setting definition line that uses `$txt['zero_to_disable']`

Sesquipedalian replied:
Yes, please restore the previous version of this line:
```
array('int', 'coppaAge', 'subtext' => $txt['zero_to_disable'], 'onchange' => 'checkCoppa();', 'onkeyup' => 'checkCoppa();'),
```

Sesquipedalian commented on Sources/ManageRegistration.php:
```diff
@@ -292,20 +322,33 @@ function SetReserved()
 	$config_vars = array(
 		array('select', 'registration_method', array($txt['setting_registration_standard'], $txt['setting_registration_activate'], $txt['setting_registration_approval'], $txt['setting_registration_disabled'])),
+		array('check', 'enableOpenID'),
+		array('check', 'notify_new_registration'),
```
Please remove these two lines. Those settings no longer exist in SMF 2.1.

Sesquipedalian commented on Sources/ManageRegistration.php:
```diff
@@ -292,20 +322,33 @@ function SetReserved()
 		array('text', 'coppaFax'),
 		array('text', 'coppaPhone'),
+	'',
+		array('check', 'announcements_default', 'disabled' => 1, 'value' => 0),
```
This setting no longer exists, because SMF 2.1 always gives the user the ability to opt out of receiving announcement emails, so please remove these two lines.

MisterVector committed:
Re-use $txt['zero_to_disable'] for coppaAge subtext

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

SychO9 commented on Sources/ManageRegistration.php:
```diff
@@ -342,7 +342,7 @@ function ModifyRegistrationSettings($return_config = false)
 		array('warning', empty($policy) ? 'error_no_privacy_policy' : ''),
 	'',
-		array('int', 'coppaAge', 'subtext' => $txt['setting_coppaAge_desc'], 'onchange' => 'checkCoppa();'),
+		array('int', 'coppaAge', 'subtext' => $txt['zero_to_disable'], 'onchange' => 'checkCoppa();'),
```
also delete the new string you added since it's no longer needed 👍

MisterVector replied:
Which string specifically?

MisterVector replied:
@SychO9 if you're referring to the `setting_coppaAge_desc` text string, that has already been removed in a different commit.

Sesquipedalian commented on Sources/ManageRegistration.php:
```diff
@@ -328,6 +371,11 @@ function ModifyRegistrationSettings($return_config = false)
+		// GDPR requires these settings to have certain values
+		$_POST['requireAgreement'] = 1;
+		$_POST['requirePolicyAgreement'] = 1;
+		$_POST['announcements_default'] = 0;
```
Hm. Simply forcing these settings to always have a certain value doesn't make a lot of sense overall. 

On the one hand, if we do decide to always force these settings to have certain values, we would do better to remove the settings entirely and simply hardcode the behaviour that they control. On the other hand, I'm not sure that removing these settings entirely is actually what we want to do, either. Some of the behavioural differences between 2.0.x and 2.1 mean that we might still need these settings even in a "GDPR compliant by default" design.

So, at least for the time being, these lines should be removed.

Sesquipedalian commented on Sources/Register.php:
```diff
@@ -209,6 +225,9 @@ function Register($reg_errors = array())
 
+	$context['announcements_ask'] = true;
+	$context['notify_announcements'] = isset($_POST['notify_announcements']) ? (bool) $_POST['notify_announcements'] : !empty($modSettings['announcements_default']);
+	
```
These lines are unneeded. `$context['announcements_ask']` doesn't exist in 2.1, and `$context['notify_announcements']` is handled elsewhere in Register.php.

Sesquipedalian commented on Sources/Register.php:
```diff
@@ -429,6 +448,9 @@ function Register2()
 	$regOptions['theme_vars'] = isset($_POST['options']) && is_array($_POST['options']) ? $_POST['options'] : array();
 
+	// Note when they accepted the agreement and privacy policy
+	$regOptions['theme_vars']['agreement_accepted'] = $regOptions['theme_vars']['policy_accepted'] = time();
```
This should be changed to:
```suggestion
	if (!empty($modSettings['requireAgreement']))
		$regOptions['theme_vars']['agreement_accepted'] = time();
	if (!empty($modSettings['requirePolicyAgreement']))
		$regOptions['theme_vars']['policy_accepted'] = time();
```
(Note to self and other dev team members: We should probably change that in 2.0.x, too.)

Sesquipedalian commented on Themes/default/Register.template.php:
```diff
@@ -166,6 +165,15 @@ function verifyAgree()
+							<input type="checkbox" name="notify_announcements" id="notify_announcements" tabindex="', $context['tabindex']++, '" class="input_check"', (!empty($context['notify_announcements']) ? ' checked="checked"' : ''), ' />
+						</dd>
+					</dl>';
+
```
This can be removed. A checkbox for this is already included here: https://github.com/SimpleMachines/SMF2.1/blob/c514281ad04c910a9afd31057dea7054e05be92d/Themes/default/Register.template.php#L137-L144

Sesquipedalian commented on Themes/default/Register.template.php:
```diff
@@ -560,6 +568,18 @@ function template_admin_register()
 						<dd>
 							<input type="checkbox" name="emailActivate" id="emailActivate_check" tabindex="', $context['tabindex']++, '"', !empty($modSettings['registration_method']) && $modSettings['registration_method'] == 1 ? ' checked' : '', ' onclick="onCheckChange();">
 						</dd>
+						<dt>
+							<strong><label for="requireAgreement">', $txt['admin_register_require_agreement'], ':</label></strong>
+						</dt>
+						<dd>
+							<input type="checkbox" name="requireAgreement" id="requireAgreement_check" tabindex="', $context['tabindex']++, '" checked="checked" disabled="disabled" class="input_check" /><input type="hidden" name="requireAgreement" value=1>
+						</dd>
+						<dt>
+							<strong><label for="requirePolicyAgreement">', $txt['admin_register_require_policy'], ':</label></strong>
+						</dt>
+						<dd>
+							<input type="checkbox" name="requirePolicyAgreement" id="requirePolicyAgreement_check" tabindex="', $context['tabindex']++, '" checked="checked" disabled="disabled" class="input_check" /><input type="hidden" name="requirePolicyAgreement" value=1>
+						</dd>
```
I think we can leave out these checkboxes in 2.1, and instead just hard code the behaviour in Register.php so that newly created users will always be presented with the agreement acceptance form when they first log in.

Sesquipedalian commented on Themes/default/languages/Errors.english.php:
```diff
@@ -426,6 +426,13 @@
+$txt['error_no_agreement'] = 'There is no registration agreement to display!';
+$txt['error_no_privacy_policy'] = 'A privacy policy has not been created for this forum.';
+
+// Unsubscribe
+$txt['unsubscribe_invalid'] = 'The unsubscribe link that brought you here does not appear to be valid.';
+
```
This can be removed. It already exists elsewhere in the file.

Sesquipedalian commented on Themes/default/languages/Help.english.php:
```diff
@@ -634,5 +634,9 @@
 $helptxt['languages_dictionary'] = 'This defines the main language dictionary that will be used by the <a href="https://php.net/function.pspell-new" target="_blank" rel="noopener">pSpell</a> spellchecker (if installed).';
 $helptxt['languages_spelling'] = 'This defines the language variant that will be used by the <a href="https://php.net/function.pspell-new" target="_blank" rel="noopener">pSpell</a> spellchecker. For many languages this is not applicable.';
 $helptxt['languages_recaptcha'] = 'The language code to use for the reCAPTCHA verification widget.<br><br>See Google\'s reCAPTCHA documentation for its <a href="https://developers.google.com/recaptcha/docs/language" target="_blank" class="bbc_link">list of supported languages</a>.';
+$helptxt['allow_disableAnnounce'] = 'This will allow users to opt out of notification of topics you announce by checking the &quot;announce topic&quot; checkbox when posting.<br /><br />This setting must be enabled in order to comply with the rules of the <a href="https://ec.europa.eu/commission/priorities/justice-and-fundamental-rights/data-protection/2018-reform-eu-data-protection-rules_en" cla ...
+$helptxt['announcements_default'] = 'This simply controls whether the the checkbox on the registration form starts as checked or unchecked.<br /><br />This setting must be disabled in order to comply with the rules of the <a href="https://ec.europa.eu/commission/priorities/justice-and-fundamental-rights/data-protection/2018-reform-eu-data-protection-rules_en" class="bbc_link">GDPR</a>.';
```
Please remove these two lines. The settings they refer to do not exist in 2.1.

Sesquipedalian commented on Themes/default/languages/Login.english.php:
```diff
@@ -91,6 +95,7 @@
 $txt['setting_coppaPost_desc'] = 'Only applies if age restriction is in place';
 $txt['setting_coppaFax'] = 'Fax number to which approval forms should be faxed';
 $txt['setting_coppaPhone'] = 'Contact number for parents to contact with age restriction queries';
+$txt['setting_announcements_default'] = 'Enable "' . $txt['notify_announcements'] . '" by default.';
```
We don't need this line in 2.1. The relevant setting already has a UI elsewhere.

Sesquipedalian commented on Themes/default/languages/Login.english.php:
```diff
@@ -107,6 +112,8 @@
 $txt['admin_register_group_desc'] = 'Primary membergroup new member will belong to';
 $txt['admin_register_group_none'] = '(no primary membergroup)';
 $txt['admin_register_done'] = 'Member %1$s has been registered successfully!';
+$txt['admin_register_require_agreement'] = 'Require user to accept the registration agreement';
+$txt['admin_register_require_policy'] = 'Require user to accept the privacy policy';
```
Removing the checkboxes in `template_admin_register()` (see above) will also mean that these lines can be removed.

Sesquipedalian commented on Themes/default/languages/Profile.english.php:
```diff
@@ -533,6 +533,12 @@
+// Registration Agreement
+$txt['trackEdit_action_agreement_accepted'] = 'Accepted the registration agreement';
+$txt['trackEdit_action_policy_accepted'] = 'Accepted the privacy policy';
+
+$txt['export_profile_data'] = 'Download profile data';
```
This can be removed. UI strings for the data export function are covered elsewhere.

Sesquipedalian reviewed (changes requested):
Thanks for your work on this, @MisterVector. Much appreciated.

This is an initial review, based only on reading through the code. Once these requested changes have been implemented, I'll run tests and then give a second review.

MisterVector committed:
Took care of a number of things at Sesquipedalian's request

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

MisterVector commented:
@Sesquipedalian you're good to go for a second review!

MisterVector commented:
@sbulen An earlier commit has addressed the `setting_coppaAge_desc` error you receive. Now as for the `log_type` issue, I'm not quite sure how to address that.

As for me, I'm getting this error:

![image](https://user-images.githubusercontent.com/437096/81025909-3578c080-8e2d-11ea-878c-386e95d1e942.png)

I'm not sure how to properly create and check the "admin-rega" token in ManageRegistration.php::EditAgreement(). I tried placing that code in respect to how the "admin-regp" token is created and checked in the EditPrivacyPolicy() function in the same file.

live627 commented:
This is what I get

( ! ) Parse error: syntax error, unexpected ''" />' (T_ENCAPSED_AND_WHITESPACE) in C:\wamp64\www\SMF2.1\Themes\default\Register.template.php on line 764

live627 commented on Themes/default/Register.template.php:
```diff
@@ -701,4 +706,72 @@ function template_edit_reserved_words()
+					</div>
+				</form>
+			</div>
+		</div>
```
yooo

MisterVector replied:
Not sure how that happened!

live627 commented on Sources/ManageRegistration.php:
```diff
@@ -292,16 +322,25 @@ function SetReserved()
-
-		array('int', 'coppaAge', 'subtext' => $txt['zero_to_disable'], 'onchange' => 'checkCoppa();', 'onkeyup' => 'checkCoppa();'),
+	'',
+		array('check', 'requireAgreement', 'text_label' => $txt['admin_agreement'], 'value' => 1, 'disabled' => 1),
+		array('warning', empty($agreement) ? 'error_no_agreement' : ''),
+		array('check', 'requirePolicyAgreement', 'text_label' => $txt['admin_privacy_policy'], 'value' => 1, 'disabled' => 1),
+		array('warning', empty($policy) ? 'error_no_privacy_policy' : ''),
```
why are these always disabled?

MisterVector replied:
Those values in SMF 2.0.16 were based on `!empty($modSettings['force_gdpr'])` but since that setting doesn't exist anymore, these values are being treated as if `force_gdpr` was always true

live627 replied:
Then you have it backwards. You can remove the disabled part I think.

MisterVector replied:
I think they wanted the agreement always required, but I might be wrong.

MisterVector replied:
@Sesquipedalian any input here?

XinYenFon replied:
Both Agreements are forced in 2.x

MisterVector replied:
Therefore I think this conversation can be resolved then.

live627 commented on Sources/Logging.php:
```diff
@@ -501,6 +501,10 @@ function logActions($logs)
 		}
 
+		$always_log = array('agreement_accepted', 'policy_accepted', 'agreement_updated', 'policy_updated');
+		if ((empty($modSettings['modlog_enabled']) && !in_array($action, $always_log)) || !isset($log_types[$log_type]))
```
Notice: Undefined variable: log_type in C:\wamp64\www\SMF2.1\Sources\Logging.php on line 505

MisterVector committed:
Not sure how this happens, but now it's fixed

live627 commented on Sources/Modlog.php:
```diff
@@ -58,9 +60,11 @@ function ViewModlog()
 			DELETE FROM {db_prefix}log_actions
-			WHERE id_log = {int:moderate_log}',
+			WHERE id_log = {int:moderate_log}
+			AND action NOT IN ({array_string:uneditable})'),
```
A parse error occurred: Syntax error, unexpected ',' on line 64 at column 50

Oldiesmann replied:
You've got an extra ) there before the comma

live627 commented on Sources/ManageSettings.php:
```diff
@@ -54,6 +54,7 @@ function ModifyFeatureSettings()
 	loadLanguage('ManageSettings');
 
 	$context['page_title'] = $txt['modSettings_title'];
+	$context['show_privacy_policy_warning'] = empty($modSettings['policy_' . $language]);
```
Missing globals: `$modSettings` and `$language`

MisterVector committed:
Fixed a few things

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

live627 commented on Sources/Modlog.php:
```diff
@@ -368,6 +375,9 @@ function list_getModLogEntries($start, $items_per_page, $sort, $query_string = '
 
 	$modlog_query = allowedTo('admin_forum') || $user_info['mod_cache']['bq'] == '1=1' ? '1=1' : (($user_info['mod_cache']['bq'] == '0=1' || $ignore_boards) ? 'lm.id_board = 0 AND lm.id_topic = 0' : (strtr($user_info['mod_cache']['bq'], array('id_board' => 'b.id_board')) . ' AND ' . strtr($user_info['mod_cache']['bq'], array('id_board' => 't.id_board'))));
 
+	if (!isset($context['uneditable_actions']))
+		$context['uneditable_actions'] = array();
```
Missing global: $context

live627 commented on Themes/default/Agreement.template.php:
```diff
@@ -0,0 +1,88 @@
+		</div>';
+
+		if ($context['can_accept_agreement'])
+		{
+			echo '
+		<div class="description">
+			', $txt['agreement_updated_desc'], '
+		</div>';
+		}
+		elseif (!empty($context['agreement_accepted_date']))
+		{
+			echo '
+		<div class="description">
+			', sprintf($txt['agreement_accepted'], timeformat($context['agreement_accepted_date'], false)), '
+		</div>';
+		}
+
+		echo '
+		<div class="roundframe">
```
I suggest both description  boxes to have the class `information noup` ans this main one `windowbg noup`

![Screenshot_2020-05-05 Privacy Policy](https://user-images.githubusercontent.com/431799/81141422-7520d480-8f21-11ea-9524-3c76438af3bb.png)

MisterVector committed:
Added missing global $context

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

XinYenFon committed:
Fixes logging related issue, updates template, small CSS extra

XinYenFon commented:
I sent a small PR (https://github.com/MisterVector/SMF2.1/pull/1) to resolve logging related issue, updated the template and added agreement.txt to gitignore. Will be listed here after @MisterVector merges it.

XinYenFon commented:
Also forgot to mention but, upgrader needs to check if user has `requirePolicyAgreement ` set & 1, user can be coming from old version of SMF where requirePolicyAgreement does not set or not even exist which creates issues (like skipping the agreement part entirely).

MisterVector commented:
Working on the upgrade scripts to set `requireAgreement` and `requirePolicyAgreement` to 1. Looking at PostgreSQL for how to replace a value, I find that `In PostgreSQL 9.5, the ON CONFLICT clause was added to INSERT`. I noticed that SMF 2.1 requires PostgreSQL 9.4, but with the addition of this statement, the version would be bumped to 9.5. Is that fine?

SychO9 commented:
@MisterVector I believe you can just do this and not worry about those details
```php
$smcFunc['db_insert']('replace',
	'{db_prefix}settings',
	array('variable' => 'string', 'value' => 'string'),
	array('requirePrivacyPolicy', '1'),
	array('variable')
);
```

However I would wait for one of the developers to confirm this is correct as I haven't dealt with the upgrader before.

MisterVector commented:
I have added what I think would work for the MySQL upgrade scripts. I haven't included the PostgreSQL scripts yet. I'll wait for word on that.

MisterVector commented:
@XinYenFon is there a reason you added agreement.txt to .gitignore when you PR'd to my pull request branch? Was that an accident?

XinYenFon commented:
it is required otherwise if someone edits their agreement it will show as new file or changed file. Ping @Sesquipedalian to fix build tools...

MisterVector commented:
@XinYenFon That's true actually, we wouldn't want people committing their license files on accident...

@Sesquipedalian probably can skip it for now, but have you thought about inserting the agreement text into the database?

MisterVector commented:
CI actually hasn't built since the first commit to this PR. Once the syntax issues were taken care of, the remaining issue has been with build tools not finding agreement.txt.

live627 committed:
is_filtered_request

live627 commented:
- `requireAgreement` is present since v1.1
- `requirePolicyAgreement` probably should not be set since SMF does not ship with a privacy policy
- The two new checkboxes added in the admin shouldn't be permanently disabled because that defeats the purpose of them
- Along the same vein, those two checkboxes always are checked, which again, defeats the purpose.

live627 committed:
don't show the agreement on popups

Sesquipedalian commented:
The build tools are fine. 🙂 The Travis CI check is failing because the license blocks in ./Sources/Agreement.php and ./Themes/default/Agreement.template.php are missing some bits. The correct format is this:
```
/**
 * <<description of file's purpose>>
 *
 * Simple Machines Forum (SMF)
 *
 * @package SMF
 * @author Simple Machines https://www.simplemachines.org
 * @copyright 2020 Simple Machines and individual contributors
 * @license https://www.simplemachines.org/about/smf/license.php BSD
 *
 * @version 2.1 RC2
 */
```
Currently, the two files are missing the URL in the author line and `and individual contributors` in the copyright line.

XinYenFon commented:
@live627 true but writing a Privacy Policy does not trigger those database values to 1 in some cases. I agree some things are logically wrong and I do think those settings should not be forced on users.

MisterVector committed:
Corrected headers of Agreement.php and Agreement.template.php

Signed-off-by: MisterVector <whizkid3000@hotmail.com>

MisterVector commented:
Wondering what is left to do. I don't yet have the queries set for the PostgreSQL upgrade scripts. @sbulen I have hopefully addressed your errors (and with the PR @XinYenFon sent, the `log_type` issue). I haven't yet figured out the error I get with `admin-rega_token` though.

Sesquipedalian commented:
I'm working on testing this and putting together some revisions that I will send to you, @MisterVector. As you may have noticed, my time has been constrained lately, so it might still be a few days yet before I can finish that. Your work is very much appreciated. :)

MisterVector commented:
Yeah I noticed you've been focusing on another important PR. Take your time man and eventually we'll get it sorted out!

MissAllSunday commented:
Hi @MisterVector The changes looks good codewise. Thanks!  the only minor change would be to change the direct checks against vars for empty() checks. This is mainly due to SMF historically prefers empty() checks over direct on var ones.

MisterVector committed:
Use empty() where applicable for direct variable checks

MisterVector committed:
Fixed syntax issue

MissAllSunday commented:
Been testing this one, so far looks good, thanks!

Been wondering if you could add an info notice similar to the one an admin gets when s/he edits the Registration Agreement, this will help in getting an indication if their changes were indeed saved.

MisterVector committed:
When editing the privacy policy, display a success message

MisterVector commented:
![image](https://user-images.githubusercontent.com/437096/89134114-201c3c80-d4d7-11ea-8c4a-5965ca7897ec.png)

@MissAllSunday please advise as this hasn't yet been added to the postgresql upgrade scripts. The install scripts have already been covered.

jdarwood007 commented on other/upgrade_2-1_mysql.sql:
```diff
@@ -51,6 +51,13 @@ CREATE TABLE IF NOT EXISTS {$db_prefix}member_logins (
+VALUES ('requireAgreement', '1'),
+	   ('requirePolicyAgreement', '1');
+---#
+
```
If you do this, they run upgrade script once, it adds it.  If they change the setting then run upgrade again, this will revert it.  We should only insert it if it doesn't exist.

If these are required and shouldn't be disabled, then it should be removed as a setting.

jdarwood007 commented on other/upgrade_1-0.sql:
```diff
@@ -1953,6 +1952,11 @@ VALUES
+REPLACE INTO {$db_prefix}settings
+	(variable, value)
+VALUES ('requireAgreement', '1'),
+	   ('requirePolicyAgreement', '1');
```
The first edit to upgrade is fine since the option is being removed, but we shouldn't otherwise touch the upgrade.

MissAllSunday commented:
> ![image](https://user-images.githubusercontent.com/437096/89134114-201c3c80-d4d7-11ea-8c4a-5965ca7897ec.png)
> 
> @MissAllSunday please advise as this hasn't yet been added to the postgresql upgrade scripts. The install scripts have already been covered.

As this is something that directly impacts the behavior of the forum as a whole, I had raised the concern to the rest of the team, once we made an agreement about it we can proceed further.

MisterVector commented:
I've had a lot of trouble installing SMF 2.1... couldn't do it on the last version of xampp that still used MySQL, and couldn't do it on the latest. Perhaps that's another thing to test for others to see if it's actually an issue for them. Perhaps I wouldn't have had an issue if I installed MySQL and PHP separately, but I don't know.

My path has typically involved installing SMF 2.0 and then later upgrading to SMF 2.1 in my PR branch.

live627 commented:
SMF works  with mariadb (fork and drop in replacement of mysql)

On Thu, Aug 6, 2020 at 1:46 PM Joel Paulien <notifications@github.com>
wrote:

> I've had a lot of trouble installing SMF 2.1... couldn't do it on the last
> version of xampp that still used MySQL, and couldn't do it on the latest.
> Perhaps that's another thing to test for others to see if it's actually an
> issue for them. Perhaps I wouldn't have had an issue if I installed MySQL
> and PHP separately, but I don't know.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6100#issuecomment-670183430>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNNZWYIUF7MVPAJMO5R3R7MJBXANCNFSM4MX7IE2A>
> .
>

MissAllSunday commented:
Hi @MisterVector The team has come to an agreement about leaving the requirePolicyAgreement setting off by default.

SychO9 commented on Sources/ManageRegistration.php:
```diff
@@ -292,16 +322,25 @@ function SetReserved()
-
-		array('int', 'coppaAge', 'subtext' => $txt['zero_to_disable'], 'onchange' => 'checkCoppa();', 'onkeyup' => 'checkCoppa();'),
+	'',
+		array('check', 'requireAgreement', 'text_label' => $txt['admin_agreement'], 'value' => 1, 'disabled' => 1),
+		array('warning', empty($agreement) ? 'error_no_agreement' : ''),
+		array('check', 'requirePolicyAgreement', 'text_label' => $txt['admin_privacy_policy'], 'value' => 1, 'disabled' => 1),
```
> leaving the requirePolicyAgreement setting off by default.

In which case, I believe this needs to no longer be disabled, to allow users enabling it.

MisterVector commented:
@SychO9 I'll go back over all the changes in this PR and see what needs to be modified.

MisterVector committed:
Don't disable the requirePolicyAgreement checkbox

MisterVector committed:
Don't disable the requireAgreement setting and set it's and
requirePolicyAgreement's values from the database settings

frandominguezl commented:
I'm currently testing this. When you click on "Sign Up", this is what you see:

![image](https://user-images.githubusercontent.com/10365222/90104631-e6ec9380-dd44-11ea-8dce-e716a88ac658.png)

After you click on "I ACCEPT THE TERMS OF THE AGREEMENT.", this is what you see:

![image](https://user-images.githubusercontent.com/10365222/90104688-fb309080-dd44-11ea-9e1f-1ed61e55351b.png)

Maybe this latter button should be placed in the first one, and change this latter to "REGISTER".

Also, if you do the change, please remove the final "." of each sentence for the buttons.

Thanks!

frandominguezl commented:
Another suggestion here. Maybe we could add some spacing between the button of "I AGREE" after changing the current privacy policy to a new one and logging into an account that needs to accept it to proceed navigation.

![image](https://user-images.githubusercontent.com/10365222/90104983-7003ca80-dd45-11ea-8970-adf1880ac5e6.png)

Thanks!

sbulen commented:
I noted two issues/questions during my testing:
 - I am not seeing user acceptance of the policies logged, maybe I'm looking in the wrong place? (in 2.0 they're in log_actions)
 - When an admin registers a user, that user is not required to accept the policies (in 2.0, the acceptance is still tracked & they will still have to agree before getting access to the forum, even if added by an admin)

Thank you for doing this!

MisterVector commented:
> 
> SMF works with mariadb (fork and drop in replacement of mysql)
> […](#)
> On Thu, Aug 6, 2020 at 1:46 PM Joel Paulien ***@***.***> wrote: I've had a lot of trouble installing SMF 2.1... couldn't do it on the last version of xampp that still used MySQL, and couldn't do it on the latest. Perhaps that's another thing to test for others to see if it's actually an issue for them. Perhaps I wouldn't have had an issue if I installed MySQL and PHP separately, but I don't know.

Previously I took the files off of master, copied them to a separate folder somewhere else, and installed SMF that way. Either I'd get an error with MariaDB when installing, or the installer would time out (odd issue).

I've just run through the same steps but instead installing off the fork itself, and everything turned out fine. So weird though!

sbulen commented:
It's a tangent here, but...  SMF does not formally support MariaDB.  There are discussions on issues over in the support forum on this.  Yes, installs sometimes have issues, due to conflicts between SMF db column names & MariaDB reserved words.  Also, their version #s have strayed, which can lead to other issues.  

You can usually work thru the issues, but it takes work...

Its a drag & drop replacement... usually...

MisterVector commented:
Hunting down another issue I found: if only the privacy policy is enabled, and a user signs up, they are not presented with the privacy policy.

MisterVector committed:
Fixed privacy policy not showing up for new user sign ups if
it was the only policy that was enabled

MisterVector committed:
User accounts created by admins must accept the user agreement
and privacy policy if they are enabled

sbulen commented:
I have confirmed that the "privacy policy only" issue was fixed.

The new user via admin account can still logon without accepting the policy, though.  There is a brief flicker on the screen like it tried to do something.

linux, php7.3

live627 commented on Sources/Logging.php:
```diff
@@ -501,6 +501,10 @@ function logActions($logs)
 		}
 
+		$always_log = array('agreement_accepted', 'policy_accepted', 'agreement_updated', 'policy_updated');
+		if ((empty($modSettings['modlog_enabled']) && !in_array($action, $always_log)) || !isset($log['log_type']))
```
`$action` doesn't exist. Should be `$log['action']`

live627 replied:
Actually... incoming PR

live627 replied:
https://github.com/MisterVector/SMF2.1/pull/3

MisterVector replied:
Does this fix the issue @sbulen reported about the agreement and policy acceptance not being logged?

live627 committed:
Fix undefined var in the action log

sbulen commented:
OK, the very last thing I see here has to do with the upgrader: 

Short version:  The changes to the 1-0, 1-1 & 2-0 upgrade sql files are not needed and should be removed.  The 2-1 upgrade file only needs to add requirePolicyAgreement, as that is the only new setting being added in 2.1 (requireAgreement has been around since 1.0).  

Longer version:  Each upgrader step takes you from the PRIOR version to THAT version.  In theory, they should never be changed on subsequent versions (rarely - only when really old bugs are found...).  If the setting didn't exist back then, we don't need to add it retroactively.  

The upgrader will literally run every upgrader step from your "FROM" version to your "TO" version (yes, including the initial "FROM" version, to ensure DB completeness).  E.g., if you are going from 1.0 to 2.1, it will execute 1.0, 1.1, 2.0 & 2.1.  So we don't add new settings in all of the scripts.  Only the script for the version where the setting was added.  

The odd hiccup here is that although the new setting was added in 2.0, it was added by the GDPR switch in the app, not in the install or upgrade scripts...  In the 2.0 code, there are no changes in 1.0, 1.1 & 2.0 upgrade sql files.  The same should be true here.  We do need to ensure that single new setting (requirePolicyAgreement) is covered in the 2.1 installs & 2.1 upgrade.

sbulen commented:
> Working on the upgrade scripts to set `requireAgreement` and `requirePolicyAgreement` to 1. Looking at PostgreSQL for how to replace a value, I find that `In PostgreSQL 9.5, the ON CONFLICT clause was added to INSERT`. I noticed that SMF 2.1 requires PostgreSQL 9.4, but with the addition of this statement, the version would be bumped to 9.5. Is that fine?

MisterVector - I just took a peek and the ON CONFLICT DO NOTHING is used throughout the 2-1 postgresql upgrade script.  So much for 9.4 compatibility...  We'll have to update our documentation.  

You should emulate one of the existing lines & you'll be fine.   

As noted in my earlier comment, you only need to worry about the new item, requirePolicyAgreement.

sbulen commented:
Upgrader changes look good - thanks!

@live627 pointed out that the ON CONFLICT DO NOTHING actually receives special attention in the upgrader for pg.  So, the 9.4 minimum version still holds.  If interested, take a peek here:
https://github.com/SimpleMachines/SMF2.1/blob/6a55d9242aecd5f42abba349b488eaada71796e4/other/upgrade.php#L2222

I'll do further testing.

Some other background you may find helpful:  The upgrader has also been used over the years to help _repair_ damaged DBs.  You should be able to re-run the upgrader on a forum without breaking it or losing settings.  

You will see in upgrade_query() that duplicate table errors, duplicate column errors, duplicate row errors, deleting a column that doesn't exist, etc., are all ignored.  HOWEVER...  Also note that, in the event of critical settings/columns missing, the upgrader can be used to fix that.  It will fill in the gaps & get the DB to where it needs to be.

sbulen commented:
@live627 I'm getting this today, hadn't seen it before.  Looks like you removed the definition & added a check for it?

![image](https://user-images.githubusercontent.com/23568484/90468340-2f9ca780-e0cb-11ea-980e-2ac4748ce09d.png)

sbulen commented:
Migrating from 2.0 to 2.1, the non-English language privacy policies, when -utf8, did not make it over.  

The problem is that the languages in 2.0 may or may not have had a -utf8 suffix.  In 2.1, they are all implicitly -utf8 already...  

Post conversion here - spanish isn't found, because it doesn't have -utf8 at the end in 2.1:
![image](https://user-images.githubusercontent.com/23568484/90470614-c91a8800-e0d0-11ea-8280-8acbd64a1fe1.png)

There are 2.1 upgrader steps to change languages to strip the -utf8 from the language strings.  I think we need one here, too.  A very safe, non-destructive one... E.g., if setting "policy_\*-utf8" exists, and "policy_\*" does NOT exist, rename, otherwise do nothing.  Remember we should be able to safely rerun it.

This would get the 'updated_' history strings as well.

Might be interesting, I don't think I've seen a setting key modified before?  A delete/add may be needed?  If you need help, let me know, I'll take a stab at it.

sbulen commented:
For that matter, the registration agreement (the *.txt file) has the same issue...  This isn't tied to this PR, so I will write that up separately.

MisterVector commented:
I won't be here for the next week, so perhaps you can take a stab at the privacy policy stuff @sbulen if you want.

sbulen commented:
Will do.

sbulen commented:
Minor question/issue: I am pretty sure the notify_tokens setting is not used in 2.1.  It is in 2.0 only.  Token based unsubscription is default behavior in 2.1, so no setting is needed.  

I saw a stray reference in there to notify_tokens that I am pretty sure must be removed???

sbulen commented:
> For that matter, the registration agreement (the *.txt file) has the same issue... This isn't tied to this PR, so I will write that up separately.

Followup note...  After further testing, I decided against writing up an issue for this.  Unlike the privacy policies, the flat file agreements are in the language packs, which need to be reinstalled after the 2.0=>2.1 upgrade.  Doing so overwrites your existing files, _including the agreement..._  Leaving them alone (with the -utf8 suffix) actually makes it easier to recover it if needed...  Renaming them increases the likelihood they're overwritten...

I think I'll leave this as-is.  If you have a custom utf8 reg agreement, you will need to rename it manually after the upgrade.  If you have a custom non-utf8 agreement, you better not overwrite it when you install the language pack...  All in all, it seems safer as-is.  

Open to input...

sbulen commented:
@live627 - have you had a chance to look at the $always_log issue above?

live627 committed:
forgot the array

live627 committed:
send it to the hook

live627 committed:
update misleading doc

sbulen commented:
Thanks @live627 !

All issues I've seen above have been addressed.  The only oddity is that policy acceptance logged in 2.0 isn't visible, but that appears to be a 2.0 issue...  Policy acceptance in 2.1 is getting logged properly & is visible under the profile edit log - but only AFTER profile edit logging is enabled.

live627 committed:
noup

live627 committed:
center these buttons

live627 commented:
oh, it should be logged regardlesss

live627 committed:
wrong operator used

live627 commented:
I have only the agreement for my user  to accept and if I use the linktree to go to the `agreement`  action, it then shows both documents.

I need to fix this  oddity...

live627 committed:
help action no longer needs to display the agreement as it is now redundant

live627 committed:
update the `terms and rules` link to the agreement section

live627 committed:
rename all `policy` context vars to `privacy_policy` for consistency

live627 commented:
Seems to me that the subaction juggling can be re moved. Unless the intent was to allow the user to view either one at will. But there's no easily accessible links for that.

sbulen commented:
I think the only need there is to do one, the other, or both.  Whatever has been updated.

live627 commented:
Yesh, manual URL overrides shouldn't be possible.

On Wed, Aug 19, 2020 at 8:15 PM sbulen <notifications@github.com> wrote:

> I think the only need there is to do one, the other, or both. Whatever has
> been updated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6100#issuecomment-676921406>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN6CKCOC47POTBRETCDSBSIMNANCNFSM4MX7IE2A>
> .
>

sbulen commented:
> oh, it should be logged regardlesss

Yes, I think you're right.

sbulen commented:
@live627 - could you take a look at these?

2 issues:
 - New user registrations are missing id_member in log_actions.  It was bugging me that I sometimes saw it blank & sometimes not.  The problem is NEW users (this is an issue in 2.0 also).  
 - This is new as of today, attempting a new user registration:

![image](https://user-images.githubusercontent.com/23568484/90820016-9695a880-e2e5-11ea-8639-36fb068fd132.png)

sbulen commented:
More info:
![image](https://user-images.githubusercontent.com/23568484/90822291-0e190700-e2e9-11ea-86d4-fc49e12db48b.png)

live627 committed:
No more janky URL  parameters

live627 committed:
wrong error strings used

live627 commented:
... this was with the profile edits log disabled, right?

Would make this much ampler if the logs could never be disabled, but hey  complexity is our old friend and people want options, right?  RIGHT? So I've been told. Developers don't know much about the human psyche. (No grudges here! Ugh...))

live627 committed:
send member_affected because it will translate to id_member

live627 committed:
disabled log types would throw undefined index errors with $always_log

live627 committed:
make it modular again

sbulen commented:
@live627, I think all the issues noted above have been dealt with:
 - No more id_log issues
 - Policy acceptance is logged whether or not the profile edit log is enabled
 - id_member is now populated upon new user registration, so the initial policy acceptance is in the log

The only oddity I am seeing now is that certain activities are flagged as having been performed by a "deleted member".  This is on all of the new user registrations:
![image](https://user-images.githubusercontent.com/23568484/90984130-fc717280-e527-11ea-8dfa-885782d68111.png)

live627 committed:
The log must also include the applicator on new registrations

sbulen commented:
@live627 - Good news! The policy is fixed.  

But I don't know how this didn't appear before, but now I'm getting the same thing for "gender"...  It's not even on my earlier screenshot above...(???):
![image](https://user-images.githubusercontent.com/23568484/90998936-18056900-e57a-11ea-9ec8-6e4e3aede5c0.png)

sbulen commented:
Note the companion PR, #6231 , now has 2 upgrader fixes related to the privacy policy:
 - strips "-utf8" from the policy setting variable names to avoid losing your policies upon migration from 2.0 =>2,1
 - fixes gaps in id_member on log_actions, correcting data caused by the 2.0 reg bug, so the initial reg & policy acceptances are visible in the logs

MisterVector commented:
You guys made a lot of progress since I went on my trip!  I’m still on vacation but figured I’d check in.  From your perspective sbulen what all is left?

live627 commented:
That error with the gender field looks like an oddity with custom   field
processing. Similar code paaths; that may be why the symptom  is similar.

On Sun, Aug 23, 2020 at 9:25 PM Joel Paulien <notifications@github.com>
wrote:

> You guys made a lot of progress since I went on my trip! I’m still on
> vacation but figured I’d check in. From your perspective sbulen what all is
> left?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6100#issuecomment-678896115>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN7NJT55MC4FCNH57SDSCHTU5ANCNFSM4MX7IE2A>
> .
>

sbulen commented:
I have just confirmed that the 'deleted member' issue on the Gender record existed prior to this PR.  I will write that up separately.

So...  **_I do think all issues have been dealt with._**  I suggest one last round of testing by everyone to confirm recent changes haven't introduced something.  

Almost there!

SychO9 commented on Sources/ManageSettings.php:
```diff
@@ -227,6 +228,7 @@ function ModifyBasicSettings($return_config = false)
 		'',
 
 		// Option-ish things... miscellaneous sorta.
+		array('check', 'notify_tokens', 'disabled' => 1),
```
what's this setting for ? it doesn't have a label in the `Features and Options` admin page, but the disabled checkbox can be seen.

sbulen replied:
Correct...  This needs to go.

It's a setting in 2.0 that doesn't apply to 2.1.  (Whether to use no-logon-necessary unsub tokens on emails...)

SychO9 commented on other/upgrade_1-0.sql:
```diff
@@ -1920,8 +1920,8 @@ upgrade_query("
 		('defaultMaxMembers', '" . (empty($GLOBALS['MembersPerPage']) ? 20 : $GLOBALS['MembersPerPage']) . "'),
 		('time_offset', '" . (empty($GLOBALS['timeoffset']) ? 0 : $GLOBALS['timeoffset']) . "'),
-		('cookieTime', '" . (empty($GLOBALS['Cookie_Length']) ? 60 : $GLOBALS['Cookie_Length']) . "'),
 		('requireAgreement', '" . @$GLOBALS['RegAgree'] . "')");
+		('cookieTime', '" . (empty($GLOBALS['Cookie_Length']) ? 60 : $GLOBALS['Cookie_Length']) . "')");
```
why move this ? I think the line 1923 would need the `;` to be changed to `,`

sbulen replied:
I think you are correct here, too...

SychO9 commented on Themes/default/Register.template.php:
```diff
@@ -300,7 +313,7 @@ function verifyAgree()
 				<input type="submit" name="accept_agreement_coppa" value="', $context['coppa_agree_below'], '" class="button">';
 	else
 		echo '
-				<input type="submit" name="regSubmit" value="', $txt['register'], '" tabindex="', $context['tabindex']++, '" class="button">';
+				<input type="submit" name="regSubmit" value="', $txt['agreement' . (!empty($context['require_policy_agreement']) ? '_policy' : '') . '_agree'], '" class="button" />';
```
I don't think we want this button editted, this is the `Signup` submit button, instead we want to edit the buttons with the name `accept_agreement` in this same template

live627 committed:
$should_log to use is_filtered_request()

live627 committed:
wrong button renamed

live627 committed:
make the no policy error behave similar to no agreement

live627 commented:
Ahh the things I do to drop eight commits...

 Totally worth it for  a cleaner branch history.

SychO9 commented on Themes/default/css/index.css:
```diff
@@ -3630,6 +3630,9 @@ h3.catbg .main_icons::before, h3.catbg .icon {
 .cat_bar + .title_bar {
 	margin-top: 0;
 }
+.cat_bar:not(:first-of-type) {
+	margin: 10px 0 0 0;
+}
```
this doesn't seem to have any effect on the the agreements page, instead it just adds extra space everywhere else, I think it should be reverted

live627 replied:
oh but it does
![Untitled](https://user-images.githubusercontent.com/431799/91364242-268a9500-e7b3-11ea-893d-b9ee9a61a6f4.png)

SychO9 replied:
no it doesn't, because the element above it has bottom margin, just try and remove these rules and you'll see

live627 replied:
huh, so it dues. SMF's highly complicated CSS is very confusing to me. At least in 2.0's CSS file I could find my way around!

SychO9 reviewed (commented):
One last thing and this should be good to be merged (the setting I mentioned previously also has still to be fixed)

MisterVector commented:
I'm back! I'll see what I can do with the remaining issues here. Thanks for helping out @live627 !

MisterVector committed:
Get rid of this, it isn't used

live627 commented:
@MisterVector np🙂

MisterVector committed:
Fixed new user sign ups

regSubmit was used in place of accept_agreement when showing the
template for new user sign ups, preventing advancing to the next
part of user registration

MisterVector commented:
The button text for just the privacy policy enabled needs to be changed:

![image](https://user-images.githubusercontent.com/437096/91368591-63a85480-e7be-11ea-8ed7-21054d37345c.png)

I'm not sure how to code the solution in the best manner possible so someone else want to take a stab at it?

Also I think I'm satisfied with testing.

live627 commented:
Will do.

Thanks for the work you put in so far!

On Wed, Aug 26, 2020 at 5:14 PM Joel Paulien <notifications@github.com>
wrote:

> The button text for just the privacy policy enabled needs to be changed:
>
> [image: image]
> <https://user-images.githubusercontent.com/437096/91368591-63a85480-e7be-11ea-8ed7-21054d37345c.png>
>
> I'm not sure how to code the solution in the best manner possible so
> someone else want to take a stab at it?
>
> Also I think I'm satisfied with testing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6100#issuecomment-681187897>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNNZMHJVW36GJFVMN4BLSCWQN7ANCNFSM4MX7IE2A>
> .
>

live627 committed:
this here style just makes a mess

live627 committed:
use correct text for when only agreeing to the privacy policy

MisterVector commented:
This was kind of fun! Sooooooo glad it's almost finished!

live627 committed:
fix coppa

live627 commented:
me too, then we can move on  with preparing the release!

On Wed, Aug 26, 2020 at 9:11 PM Joel Paulien <notifications@github.com>
wrote:

> This was kind of fun! Sooooooo glad it's almost finished!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/6100#issuecomment-681370551>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADJNN63BXEUQ56XBKCOG5LSCXMIFANCNFSM4MX7IE2A>
> .
>

SychO9 reviewed (approved):
I believe this is finally good to go.

Thank you @MisterVector !

MisterVector committed:
Removed extra space in input name

sbulen commented:
I retested everything & it looks good!

MissAllSunday commented:
Thank you everyone for your efforts!
